### PR TITLE
Gptmd approach update

### PR DESCRIPTION
**Pull request by trishorts**
GPTMD is promiscuous in the addition of potential modifications to the xml database. This PR reduces the number of candidate modifications added to those that produce the highest score for each possible PTM. The high level details of the new algorithm are as follows:
1. Perform notch based search
2. Find modifications that match each notch
3. Find motifs for each modification and create a corresponding peptideWithSetModifications
4. Fragment each peptideWithSetModifications and compute the MetaMorpheus score.
5. Choose a subset of peptideWithSetMods having the highest score and add those localized modifications to the new xml

For bottom up:
six mann A549 files with human fasta.
old method added 200513 mods; new method added 128449 mods
old method 102324 psms; new 103546
old 39283 peptides; new 39277
old 6042 proteins; new 6012

For top down:
14 fractions x 2 techreps jurkate td files from sean dai paper
old method added 19188 mods; new method added 11013 mods
old method 23688 psms; new 24022
old 904 proteoforms; new 899
old 279 proteins; new 273

Additonal updates:
1. Eliminated output of candidate psms in .psmtsv
2. GPTMD database created with parallelization
3. Eliminated PEP from the FDR analysis

### CODE REVIEW

File: MetaMorpheus/Test/ProteinGroupTest.cs [hunk 1]
```diff
@@ -217,13 +217,13 @@ public static void TestModificationInfoListInProteinGroupsOutput()
             int totalNumberOfMods = proteins.Sum(p => p.OneBasedPossibleLocalizedModifications.Count + p.SequenceVariations.Sum(sv => sv.OneBasedModifications.Count));
 
             //tests that modifications are being done correctly
-            Assert.AreEqual(0, totalNumberOfMods);
```
Comment (nbollis) [inline MetaMorpheus/Test/ProteinGroupTest.cs:220]: The change of an expected value with the comment above, "test that mods are being done correctly" is worrisome

**Review by Alexander-Sol — APPROVED** [review]

File: MetaMorpheus/TaskLayer/GPTMDTask/GPTMDTask.cs [hunk 2]
```diff
@@ -62,18 +79,13 @@ protected override MyTaskResults RunSpecific(string OutputFolder, List<DbForTask
             ProseCreatedWhileRunning.Append("precursor mass tolerance(s) = {" + tempSearchMode.ToProseString() + "}; ");
 
             ProseCreatedWhileRunning.Append("product mass tolerance = " + CommonParameters.ProductMassTolerance + ". ");
-            ProseCreatedWhileRunning.Append("The combined search database contained " + proteinList.Count(p => !p.IsDecoy) + " non-decoy protein entries including " + proteinList.Where(p => p.IsContaminant).Count() + " contaminant sequences. ");
```
[the remark this answers is not in the document]
Comment (trishorts) [reply MetaMorpheus/TaskLayer/GPTMDTask/GPTMDTask.cs:65]: nope. i put it back.

File: MetaMorpheus/Test/ProteinGroupTest.cs (hunk 1 above)
Comment (trishorts) [reply MetaMorpheus/Test/ProteinGroupTest.cs:219]: idk it is 8 now. maybe this is old news.

File: MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs [hunk 3]
```diff
@@ -65,105 +63,188 @@ public static bool ModFits(Modification attemptToLocalize, IBioPolymer protein,
...
+                                    if (newPeptides.Any())
                                     {
-                                        bool modIsBeforeVariant = indexInProtein < variant.OneBasedBeginPosition + offset;
-                                        bool modIsOnVariant = variant.OneBasedBeginPosition + offset <= indexInProtein && indexInProtein <= variant.OneBasedEndPosition + offset;
+                                        var scores = new List<double>();
+                                        var targetFragmentsForEachDissociationType = new ConcurrentDictionary<DissociationType, List<Product>>();
 
-                                        // if a variant protein and the mod is on the variant, index to the variant protein sequence
-                                        if (modIsOnVariant)
+                                        if (CommonParameters.DissociationType == DissociationType.Autodetect)
```
[the remark this answers is not in the document]
Comment (trishorts) [reply MetaMorpheus/EngineLayer/Gptmd/GptmdEngine.cs]: eliminated as unused code

**Review by nbollis — APPROVED** [review]

**Review by Alexander-Sol — APPROVED** [review]

**Pull request merged**